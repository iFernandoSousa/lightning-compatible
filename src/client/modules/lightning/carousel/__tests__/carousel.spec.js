/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { createElement } from 'lwc';
import { shadowQuerySelector, querySelectorAll } from 'lightning/testUtils';
import Element from 'lightningtest/carouselTest';

const createCarousel = attributes => {
    const element = createElement('lightningtest-carousel-test', {
        is: Element
    });

    Object.assign(element, attributes);
    document.body.appendChild(element);

    return element;
};

const DEFAULT_ITEMS = [
    {
        alternativeText: 'https://www.salesforce.com',
        description: 'This is a card',
        header: 'First card',
        href: 'http://www.salesforce.com',
        src:
            'https://latest-212.lightningdesignsystem.com/assets/images/carousel/carousel-01.jpg'
    },

    {
        alternativeText: 'https://www.salesforce.com',
        description: 'This is a card',
        header: 'Second card',
        href: 'http://www.salesforce.com',
        src:
            'https://latest-212.lightningdesignsystem.com/assets/images/carousel/carousel-02.jpg'
    },

    {
        alternativeText: 'https://www.salesforce.com',
        description: 'This is a card',
        header: 'Third card',
        href: 'http://www.salesforce.com',
        src:
            'https://latest-212.lightningdesignsystem.com/assets/images/carousel/carousel-03.jpg'
    }
];

function assertNthItemActive(n) {
    return element => {
        const panels = querySelectorAll(element, '.slds-carousel__panel');
        const indicators = querySelectorAll(
            element,
            '.slds-carousel__indicator'
        );

        expect(panels).toHaveLength(indicators.length);

        panels.forEach((panel, index) => {
            const tabpanel = shadowQuerySelector(panel, '[role=tabpanel]');
            expect(tabpanel.getAttribute('aria-hidden')).toBe(
                index === n ? 'false' : 'true'
            );
        });
    };
}

const assertFirstItemActive = assertNthItemActive(0);
const assertSecondItemActive = assertNthItemActive(1);
const assertThirdItemActive = assertNthItemActive(2);

describe('c-carousel', () => {
    it('carousel with 3 images', () => {
        const element = createCarousel({ items: DEFAULT_ITEMS });

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    describe('auto-scroll', () => {
        jest.useFakeTimers();

        it('first item active initially', () => {
            const element = createCarousel({ items: DEFAULT_ITEMS });

            return Promise.resolve().then(() => {
                assertFirstItemActive(element);
            });
        });

        it('second item active after first timer', () => {
            const element = createCarousel({ items: DEFAULT_ITEMS });

            return Promise.resolve().then(() => {
                jest.runOnlyPendingTimers();

                return Promise.resolve().then(() => {
                    assertSecondItemActive(element);
                });
            });
        });

        it('third item active after second timer', () => {
            const element = createCarousel({ items: DEFAULT_ITEMS });

            return Promise.resolve().then(() => {
                jest.runOnlyPendingTimers();
                return Promise.resolve().then(() => {
                    jest.runOnlyPendingTimers();
                    return Promise.resolve().then(() => {
                        assertThirdItemActive(element);
                    });
                });
            });
        });
    });

    describe('auto-scroll with auto-refresh', () => {
        jest.useFakeTimers();

        it('first item active after third timer', () => {
            const element = createCarousel({ items: DEFAULT_ITEMS });

            return Promise.resolve().then(() => {
                jest.runOnlyPendingTimers();
                return Promise.resolve().then(() => {
                    jest.runOnlyPendingTimers();
                    return Promise.resolve().then(() => {
                        jest.runOnlyPendingTimers();
                        return Promise.resolve().then(() => {
                            assertFirstItemActive(element);
                        });
                    });
                });
            });
        });
    });

    describe('auto-scroll without auto-refresh', () => {
        jest.useFakeTimers();

        it('third item active after third timer', () => {
            const element = createCarousel({
                items: DEFAULT_ITEMS,
                disableAutoRefresh: true
            });

            return Promise.resolve().then(() => {
                jest.runOnlyPendingTimers();
                return Promise.resolve().then(() => {
                    jest.runOnlyPendingTimers();
                    return Promise.resolve().then(() => {
                        jest.runOnlyPendingTimers();
                        return Promise.resolve().then(() => {
                            assertThirdItemActive(element);
                        });
                    });
                });
            });
        });
    });
    it('binds the right aria-control', () => {
        const element = createCarousel({ items: DEFAULT_ITEMS });

        return Promise.resolve().then(() => {
            const carousel = element.shadowRoot.querySelector('c-carousel');

            const paginationItem = carousel.shadowRoot.querySelector(
                '.slds-carousel__indicator a'
            );

            const ariaControls = paginationItem.getAttribute('aria-controls');
            const firstImage = carousel
                .querySelector('c-carousel-image')
                .shadowRoot.querySelector('div');
            const firstImageId = firstImage.id;

            expect(ariaControls).toEqual(firstImageId);
        });
    });

    it('binds the right aria-describedby', () => {
        const element = createCarousel({ items: DEFAULT_ITEMS });

        return Promise.resolve().then(() => {
            const carousel = element.shadowRoot.querySelector('c-carousel');

            const paginationItem = carousel.shadowRoot.querySelector(
                '.slds-carousel__indicator a'
            );

            const firstPaginationElementId = paginationItem.id;
            const firstImage = carousel
                .querySelector('c-carousel-image')
                .shadowRoot.querySelector('div');
            const firstImageAriaLablledBy = firstImage.getAttribute(
                'aria-labelledby'
            );

            expect(firstPaginationElementId).toEqual(firstImageAriaLablledBy);
        });
    });
});
