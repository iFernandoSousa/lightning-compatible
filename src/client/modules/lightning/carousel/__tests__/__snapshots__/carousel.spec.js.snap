// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-carousel carousel with 3 images 1`] = `
<lightningtest-carousel-test>
  #shadow-root(open)
    <c-carousel
      data-handles-touch="true"
    >
      #shadow-root(open)
        <div
          class="slds-carousel"
        >
          <div
            class="slds-carousel__stage"
          >
            <span
              class="slds-carousel__autoplay"
            >
              <button
                aria-pressed="false"
                class="slds-button slds-button_icon slds-button_icon-border-filled slds-button_icon-x-small"
                title="Stop auto-play"
              >
                <c-primitive-icon
                  class="slds-button__icon"
                >
                  #shadow-root(open)
                    <svg
                      aria-hidden="true"
                      class="slds-icon slds-icon-text-default slds-icon_x-small"
                      data-key="pause"
                      focusable="false"
                    >
                      <use
                        xlink:href="/assets/icons/utility-sprite/svg/symbols.svg#pause"
                      />
                    </svg>
                </c-primitive-icon>
                <span
                  class="slds-assistive-text"
                >
                  Stop auto-play
                </span>
              </button>
            </span>
            <div
              class="slds-carousel__panels"
            >
              <slot />
            </div>
            <ul
              class="slds-carousel__indicators"
              role="tablist"
            >
              <li
                class="slds-carousel__indicator"
                data-index="0"
                role="presentation"
              >
                <a
                  aria-controls="[shadow:guid]"
                  aria-selected="true"
                  class="slds-carousel__indicator-action slds-is-active"
                  href="javascript:void(0);"
                  id="[shadow:guid]"
                  role="tab"
                  tabindex="0"
                  title="This is a card Tab"
                >
                  <span
                    class="slds-assistive-text"
                  >
                    This is a card Tab
                  </span>
                </a>
              </li>
              <li
                class="slds-carousel__indicator"
                data-index="1"
                role="presentation"
              >
                <a
                  aria-controls="[shadow:guid]"
                  aria-selected="false"
                  class="slds-carousel__indicator-action"
                  href="javascript:void(0);"
                  id="[shadow:guid]"
                  role="tab"
                  tabindex="-1"
                  title="This is a card Tab"
                >
                  <span
                    class="slds-assistive-text"
                  >
                    This is a card Tab
                  </span>
                </a>
              </li>
              <li
                class="slds-carousel__indicator"
                data-index="2"
                role="presentation"
              >
                <a
                  aria-controls="[shadow:guid]"
                  aria-selected="false"
                  class="slds-carousel__indicator-action"
                  href="javascript:void(0);"
                  id="[shadow:guid]"
                  role="tab"
                  tabindex="-1"
                  title="This is a card Tab"
                >
                  <span
                    class="slds-assistive-text"
                  >
                    This is a card Tab
                  </span>
                </a>
              </li>
            </ul>
          </div>
        </div>
      <c-carousel-image
        class="slds-carousel__panel"
        data-handles-touch="true"
      >
        #shadow-root(open)
          <div
            aria-hidden="false"
            aria-labelledby="[shadow:guid]"
            id="[shadow:guid]"
            role="tabpanel"
          >
            <a
              class="slds-carousel__panel-action slds-text-link_reset"
              href="http://www.salesforce.com"
              tabindex="0"
            >
              <div
                class="slds-carousel__image"
              >
                <img
                  alt="https://www.salesforce.com"
                  src="https://latest-212.lightningdesignsystem.com/assets/images/carousel/carousel-01.jpg"
                />
              </div>
              <div
                class="slds-carousel__content"
              >
                <h2
                  class="slds-carousel__content-title"
                >
                  First card
                </h2>
                <p>
                  This is a card
                </p>
              </div>
            </a>
          </div>
      </c-carousel-image>
      <c-carousel-image
        class="slds-carousel__panel"
        data-handles-touch="true"
      >
        #shadow-root(open)
          <div
            aria-hidden="true"
            aria-labelledby="[shadow:guid]"
            id="[shadow:guid]"
            role="tabpanel"
          >
            <a
              class="slds-carousel__panel-action slds-text-link_reset"
              href="http://www.salesforce.com"
              tabindex="-1"
            >
              <div
                class="slds-carousel__image"
              >
                <img
                  alt="https://www.salesforce.com"
                  src="https://latest-212.lightningdesignsystem.com/assets/images/carousel/carousel-02.jpg"
                />
              </div>
              <div
                class="slds-carousel__content"
              >
                <h2
                  class="slds-carousel__content-title"
                >
                  Second card
                </h2>
                <p>
                  This is a card
                </p>
              </div>
            </a>
          </div>
      </c-carousel-image>
      <c-carousel-image
        class="slds-carousel__panel"
        data-handles-touch="true"
      >
        #shadow-root(open)
          <div
            aria-hidden="true"
            aria-labelledby="[shadow:guid]"
            id="[shadow:guid]"
            role="tabpanel"
          >
            <a
              class="slds-carousel__panel-action slds-text-link_reset"
              href="http://www.salesforce.com"
              tabindex="-1"
            >
              <div
                class="slds-carousel__image"
              >
                <img
                  alt="https://www.salesforce.com"
                  src="https://latest-212.lightningdesignsystem.com/assets/images/carousel/carousel-03.jpg"
                />
              </div>
              <div
                class="slds-carousel__content"
              >
                <h2
                  class="slds-carousel__content-title"
                >
                  Third card
                </h2>
                <p>
                  This is a card
                </p>
              </div>
            </a>
          </div>
      </c-carousel-image>
    </c-carousel>
</lightningtest-carousel-test>
`;
